File: Application/Loans/LoanService.cs
using System.Net;
using Application.Abstractions;
using Application.Common;
using Domain.Entities;
using Domain.Enums;

namespace Application.Loans;

public sealed class LoanService(
    ILoanRepository loans,
    IUserRepository users,
    ILoanStatusCache loanStatusCache,
    IUnitOfWork unitOfWork) : ILoanService
{
    public async Task<Result<IReadOnlyList<LoanResponse>>> GetAllAsync(CancellationToken cancellationToken)
        => Result<IReadOnlyList<LoanResponse>>.Response(
            (await loans.GetAllAsync(cancellationToken)).Select(ToResponse).ToList(),
            "Loans retrieved successfully.",
            HttpStatusCode.OK);

    public async Task<Result<IReadOnlyList<LoanResponse>>> GetByUserIdAsync(Guid userId, CancellationToken cancellationToken)
        => Result<IReadOnlyList<LoanResponse>>.Response(
            (await loans.GetByUserIdAsync(userId, cancellationToken)).Select(ToResponse).ToList(),
            "Loans retrieved successfully.",
            HttpStatusCode.OK);

    public async Task<Result<LoanResponse>> GetByIdAsync(Guid id, CancellationToken cancellationToken)
    {
        var loan = await loans.GetByIdAsync(id, cancellationToken);
        return loan is null
            ? Result<LoanResponse>.Failure("LOAN_NOT_FOUND", "Loan was not found.", HttpStatusCode.BadRequest)
            : Result<LoanResponse>.Response(ToResponse(loan), "Loan retrieved successfully.", HttpStatusCode.OK);
    }

    public async Task<Result<object>> GetStatusAsync(Guid id, CancellationToken cancellationToken)
    {
        var status = await loanStatusCache.GetOrCreateAsync(
            id,
            () => loans.GetStatusAsync(id, cancellationToken));

        if (!status.HasValue)
            return Result<object>.Failure("LOAN_NOT_FOUND", "Loan was not found.", HttpStatusCode.BadRequest);

        return Result<object>.Response(
            new { LoanId = id, Status = status.Value },
            "Loan status retrieved successfully.",
            HttpStatusCode.OK);
    }

    public async Task<Result<LoanResponse>> CreateAsync(CreateLoanRequest request, CancellationToken cancellationToken)
    {
        var user = await users.GetByIdAsync(request.UserId, cancellationToken);
        if (user is null || !user.IsActive)
            return Result<LoanResponse>.Failure(
                "LOAN_USER_NOT_FOUND",
                "Loan user was not found or is inactive.",
                HttpStatusCode.BadRequest);

        var loan = Loan.Request(request.UserId, request.Amount, request.TermInMonths, request.Purpose);
        loans.Add(loan);
        await unitOfWork.SaveChangesAsync(cancellationToken);
        loanStatusCache.Set(loan.Id, loan.Status);

        return Result<LoanResponse>.Response(ToResponse(loan), "Loan requested successfully.", HttpStatusCode.Created);
    }

    public async Task<Result<LoanResponse>> UpdateAsync(Guid id, UpdateLoanRequest request, CancellationToken cancellationToken)
    {
        var loan = await loans.GetByIdAsync(id, cancellationToken);
        if (loan is null)
            return Result<LoanResponse>.Failure("LOAN_NOT_FOUND", "Loan was not found.", HttpStatusCode.BadRequest);

        loan.UpdateRequest(request.Amount, request.TermInMonths, request.Purpose);
        await unitOfWork.SaveChangesAsync(cancellationToken);
        loanStatusCache.Set(loan.Id, loan.Status);

        return Result<LoanResponse>.Response(ToResponse(loan), "Loan updated successfully.", HttpStatusCode.OK);
    }

    public async Task<Result<LoanResponse>> ApproveAsync(Guid id, Guid adminUserId, CancellationToken cancellationToken)
    {
        Loan? reviewedLoan = null;

        await unitOfWork.ExecuteInTransactionAsync(async ct =>
        {
            var loan = await loans.GetByIdAsync(id, ct);
            if (loan is null)
                return;

            loan.Approve(adminUserId);
            await unitOfWork.SaveChangesAsync(ct);
            reviewedLoan = loan;
        }, cancellationToken);

        if (reviewedLoan is null)
            return Result<LoanResponse>.Failure("LOAN_NOT_FOUND", "Loan was not found.", HttpStatusCode.BadRequest);

        loanStatusCache.Set(id, LoanStatus.Approved);
        return Result<LoanResponse>.Response(ToResponse(reviewedLoan!), "Loan approved successfully.", HttpStatusCode.OK);
    }

    public async Task<Result<LoanResponse>> RejectAsync(
        Guid id,
        Guid adminUserId,
        RejectLoanRequest request,
        CancellationToken cancellationToken)
    {
        Loan? reviewedLoan = null;

        await unitOfWork.ExecuteInTransactionAsync(async ct =>
        {
            var loan = await loans.GetByIdAsync(id, ct);
            if (loan is null)
                return;

            loan.Reject(adminUserId, request.Reason);
            await unitOfWork.SaveChangesAsync(ct);
            reviewedLoan = loan;
        }, cancellationToken);

        if (reviewedLoan is null)
            return Result<LoanResponse>.Failure("LOAN_NOT_FOUND", "Loan was not found.", HttpStatusCode.BadRequest);

        loanStatusCache.Set(id, LoanStatus.Rejected);
        return Result<LoanResponse>.Response(ToResponse(reviewedLoan!), "Loan rejected successfully.", HttpStatusCode.OK);
    }

    public async Task<Result<object>> DeleteAsync(Guid id, CancellationToken cancellationToken)
    {
        var loan = await loans.GetByIdAsync(id, cancellationToken);
        if (loan is null)
            return Result<object>.Failure("LOAN_NOT_FOUND", "Loan was not found.", HttpStatusCode.BadRequest);

        if (loan.Status != LoanStatus.Pending)
            return Result<object>.Failure(
                "LOAN_DELETE_INVALID_STATUS",
                "Only pending loans can be deleted.",
                HttpStatusCode.BadRequest);

        loans.Remove(loan);
        await unitOfWork.SaveChangesAsync(cancellationToken);
        loanStatusCache.Remove(id);

        return Result<object>.Response("Loan deleted successfully.", HttpStatusCode.OK);
    }

    private static LoanResponse ToResponse(Loan loan)
        => new(
            loan.Id,
            loan.UserId,
            loan.Amount.Amount,
            loan.Term.Months,
            loan.Purpose.Value,
            loan.Status,
            loan.ReviewedByUserId,
            loan.ReviewedAt,
            loan.RejectionReason?.Value);
}

File: client/src/services/loanService.ts
import { apiRequest } from "./apiClient";
import type { ApiResult, CreateLoanRequest, Loan, LoanStatus, RawLoan, UpdateLoanRequest } from "../types/api";

const statusByNumber: Record<number, LoanStatus> = {
  1: "Pending",
  2: "Approved",
  3: "Rejected"
};

const normalizeLoanStatus = (status: RawLoan["status"]): LoanStatus => {
  if (typeof status === "number") {
    return statusByNumber[status] ?? "Pending";
  }

  return status;
};

const normalizeLoan = (loan: RawLoan): Loan => ({
  ...loan,
  status: normalizeLoanStatus(loan.status)
});

const normalizeLoanResult = (result: ApiResult<RawLoan>): ApiResult<Loan> => ({
  ...result,
  entity: result.entity ? normalizeLoan(result.entity) : null
});

const normalizeLoanListResult = (result: ApiResult<RawLoan[]>): ApiResult<Loan[]> => ({
  ...result,
  entity: result.entity?.map(normalizeLoan) ?? null
});

export const loanService = {
  list: async () => normalizeLoanListResult(await apiRequest<RawLoan[]>("/api/loans")),
  get: async (id: string) => normalizeLoanResult(await apiRequest<RawLoan>(`/api/loans/${id}`)),
  create: (request: CreateLoanRequest) =>
    apiRequest<RawLoan>("/api/loans", {
      method: "POST",
      body: request
    }).then(normalizeLoanResult),
  update: (id: string, request: UpdateLoanRequest) =>
    apiRequest<RawLoan>(`/api/loans/${id}`, {
      method: "PUT",
      body: request
    }).then(normalizeLoanResult),
  approve: (id: string) =>
    apiRequest<RawLoan>(`/api/loans/${id}/approve`, {
      method: "POST"
    }).then(normalizeLoanResult),
  reject: (id: string, reason: string) =>
    apiRequest<RawLoan>(`/api/loans/${id}/reject`, {
      method: "POST",
      body: { reason }
    }).then(normalizeLoanResult),
  remove: (id: string) =>
    apiRequest<object>(`/api/loans/${id}`, {
      method: "DELETE"
    })
};
